src/CpuAffinitySet.cc

```diff
@@ -52,19 +56,27 @@ CpuAffinitySet::apply()
     }
     if (!success)
         CPU_ZERO(&theOrigCpuSet);
+
+    if (!reconfiguring)
+        leave_suid();
 }
 
 void
 CpuAffinitySet::undo()
 {
     if (applied()) {
+        // TODO: use CAP_SYS_NICE capability instead
+        if (!reconfiguring)
+            enter_suid();
         if (sched_setaffinity(0, sizeof(theOrigCpuSet), &theOrigCpuSet)) {
             int xerrno = errno;
             debugs(54, DBG_IMPORTANT, "ERROR: failed to restore original CPU "
                    "affinity for process PID " << getpid() << ": " <<
                    xstrerr(xerrno));
         }
         CPU_ZERO(&theOrigCpuSet);
+        if (!reconfiguring)
+            leave_suid();
```

Comment:
> PR commit 7ec84c528813c745c9896070dd2764d986036cbe: Restrict SUID changes to startup
> reconfigure runner already operates with enter_suid() context

I am not against CpuAffinitySet requiring an suid context, either explicitly (i.e. by calling enter/exit_suid()) or implicitly (i.e. by requiring the higher level code to provide that context for CpuAffinitySet), but CpuAffinitySet should not know anything about "runners" and their relationship with suid contexts, especially when that relationship is undocumented.

* The correct long-term solution for this part of the problem is to make nested enter/exit_suid() calls a no-op (among other suid scope control improvements). That is a difficult project that deserves its own PR. I assume you do not want to block this PR on that future project, but it is your call.
* AFAICT, the correct medium-term solution for this part of the problem is to move these explicit enter/exit_suid() calls (without a `reconfiguring` guard) to useConfig() if (and only if) useConfig() is not executed inside an suid context already. Otherwise, that solution is to remove these explicit calls as unnecessary.

P.S. There is no "reconfigure runner". There are two event handlers dedicated to two different reconfiguration stages/events.

Reply:
> I am not against CpuAffinitySet requiring an suid context, either explicitly (i.e. by calling enter/exit_suid()) or implicitly (i.e. by requiring the higher level code to provide that context for CpuAffinitySet), but CpuAffinitySet should not know anything about "runners" and their relationship with suid contexts, especially when that relationship is undocumented.
> 

This is not adding any knowledge of runners. These ```if (!reconfiguring)``` checks are preserving current behaviour of this CpuAffinitySet code being run with low-privileges during reconfigure signal handling. I assume that behaviour was intentional. Otherwise the if-conditions should be dropped.


>     * The correct long-term solution for this part of the problem is to make nested enter/exit_suid() calls a no-op (among other suid scope control improvements). That is a difficult project that deserves its own PR. I assume you do not want to block this PR on that future project, but it is your call.

IMO the correct solution is to follow security best practice when dealing with SUID. That is: to minimize the amount of time spent with high privileges.

For purposes of this PR we are moving this code from a codepath with high privileges to one providing low privileges. Thus we need to add these calls around (only) the syscall which requires high privilege.

> 
>     * AFAICT, the correct medium-term solution for this part of the problem is to move these explicit enter/exit_suid() calls (without a `reconfiguring` guard) to useConfig()

Absolutely not.

The use of Must() exceptions for validation in the CpuAffinitySet code and indirectly the memory allocators it uses, would open a window for the leave_suid() code to be bypassed. Squid would return to normal operations with high SUID privileges after the exception handlers.


> if (and only if) useConfig() is not executed inside an suid context already. Otherwise, that solution is to remove these explicit calls as unnecessary.
> 

I have audited the Runners events and they all (except bootstrapConfig) run in the leave_suid() privilege level. So according to your information about ```sched_set*()```requiring root these are necessary (at least on startup/shutdown).

Reply:
> > I am not against CpuAffinitySet requiring an suid context, but CpuAffinitySet should not know anything about "runners" and their relationship with suid contexts, especially when that relationship is undocumented.

> This is not adding any knowledge of runners.

The proposed code relies on the assumption that the useConfig()/initialization event currently runs outside an suid context and the syncConfig()/reconfiguration event currently runs inside an suid context.

> These `if (!reconfiguring)` checks are preserving current behaviour of this CpuAffinitySet code being run with low-privileges during reconfigure signal handling. I assume that behaviour was intentional. Otherwise the if-conditions should be dropped.

My change request is not about whether this code preserves the current behavior and is not about whether that current behavior is correct. Those are separate issues.

> > * The correct long-term solution for this part of the problem is to make nested enter/exit_suid() calls a no-op (among other suid scope control improvements). That is a difficult project that deserves its own PR. I assume you do not want to block this PR on that future project, but it is your call.
 
> IMO the correct solution is to follow security best practice when dealing with SUID. That is: to minimize the amount of time spent with high privileges.

That is not a solution but a principle. I agree with that principle, and the long-term solution I outlined does not contradict it. In fact, it makes it a lot easier and safer to follow that principle!

Unfortunately, Squid code is not ready to fully follow that principle yet. PR changes also do not fully follow the "minimize the amount of time spent with high privileges" principle, but that is OK -- no reasonable in-scope solution would be able to do that now. PR changes make things worse in other areas, even though they _are_ inspired by the correct principle.  There is an overall better way to handle this situation given the current code limitations.

> For purposes of this PR we are moving this code from a codepath with high privileges to one providing low privileges. Thus we need to add these calls around (only) the syscall which requires high privilege.

This PR does not move this code to a low-privileged section. If it did, these `if` guards would be unnecessary! What actually happens in the current PR code is that the same code runs in two different sections (depending on the event): one that has high privileges and one that has low. That dual placement requires special handing such as these `if` guards. However, these guards create problems for future code improvements.

The medium-term solution I am proposing does not attempt to solve all the problems, but it does not make things worse. It is a uncontroversial step forward. Maybe not as big of a step as we want, but the suid management code is just not ready for bigger steps.

> > * AFAICT, the correct medium-term solution for this part of the problem is to move these explicit enter/exit_suid() calls (without a `reconfiguring` guard) to useConfig()
 
> Absolutely not.
>
> The use of Must() exceptions for validation in the CpuAffinitySet code and indirectly the memory allocators it uses, would open a window for the leave_suid() code to be bypassed. Squid would return to normal operations with high SUID privileges after the exception handlers.

That danger is already present in the official code, and even the current PR code is not completely immune to it! We do not have to address these old/existing problems in _this_ PR. However, if you do want to address them here, you can still follow my medium-term suggestion _and_ add a try/catch block inside the new enter/exit_suid() calls pair.

Reply:
You are making this situation far more complicated than it needs to be. None of the extra refactoring or handling is necessary if we just accept that the PR current code is better than current official code.

The most likely future changes to this code are someone implementing CAP_SYS_NICE and removing these SUID changes entirely. OR, properly testing the affinity reconfigure behaviour to determine whether the if-guards are necessary.


> It is a uncontroversial step forward.

Disagree. It is controversial as evidenced by our disagreements on security aspects.

> Maybe not as big of a step as we want, but the suid management code is just not ready for bigger steps.

IMO the proposed half-step is worse than the half-step this PR already implements.

Reply:
> None of the extra refactoring or handling is necessary if we just accept that the PR current code is better than current official code.

I disagree that the current PR code is better than the official code. PR changes make the code (subject to this change request) worse, not better. Yes, some work is necessary to address this problem (for the time being). I believe that work is not very complex -- no SUID refactoring is needed to find a good-enough solution (while we wait for SUID refactoring which is a big project indeed). If the amount of work is what stops you from moving this PR forward, I can volunteer to do this work.

Reply:
> Alex: CpuAffinityReconfigure() in the official code runs with the same privileges as syncConfig() used by the PR code. 

> Amos: I have audited the Runners events and they all (except bootstrapConfig) run in the leave_suid() privilege level. So according to your information about sched_set*()requiring root these are necessary (at least on startup/shutdown).

AFAICT, the combination of the above statements implies that the PR code should call enter/exit_suid() _unconditionally_, addressing this change request. Do you see a problem with that solution?